int64 p
int64 i
int64 d
